### DISCUSSION TITLE
Remark42 without a subdomain (remark.css - 404)

### DISCUSSION THREAD

**Discussion by Ivan-Chupin** [umputun/remark42#967] [Q&A]
![remark css 404](https://user-images.githubusercontent.com/83099323/116794078-edde5080-aae3-11eb-992f-2a5384352d98.png)

**Comment by umputun** [comment]
this information is not sufficient to answer anything reasonable. As this is not a bug report moved to discussions

**Comment by Ivan-Chupin** [comment]
The subdomain configuration works fine. Using Remark42 without a subdomain has several advantages. When trying to apply "[How to configure remark42 without a subdomain](https://github.com/umputun/remark42/wiki/How-to-configure-remark42-without-a-subdomain)", the remark.css request from iframe.html returns a 404 error.

**Comment by Ivan-Chupin** [comment]
```Nginx
user nginx;
pid /var/run/nginx.pid;
worker_processes auto;
worker_rlimit_nofile 8192;
error_log /var/log/nginx/error.log warn;
events {worker_connections 4096;}

http {
	sendfile on;
	tcp_nopush on;
	include /etc/nginx/mime.types;
	charset UTF-8;
	ssl_protocols TLSv1.2 TLSv1.3;
	ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384;
	ssl_session_tickets on;
	ssl_session_cache shared:SSL:10m;
	ssl_session_timeout 1d;
	ssl_stapling on;
	ssl_stapling_verify on;
	ssl_buffer_size 4k;
	ssl_early_data on;
	resolver 1.1.1.1 1.0.0.1 8.8.8.8 8.8.4.4;
	resolver_timeout 2s;
	access_log off;
	log_not_found off;
	gzip_vary on;
	gzip_proxied any;

	server {
		listen 80;
		listen [::]:80;
		server_name domain.com www.domain.com;
		return 301 https://domain.com$request_uri;
	}

	server {
		listen 443 ssl http2;
		listen [::]:443 ssl http2;
		ssl_certificate /etc/nginx/ssl/domain.com.cert.pem;
		ssl_certificate_key /etc/nginx/ssl/domain.com.key.pem;
		add_header X-Frame-Options SAMEORIGIN;
		add_header X-Content-Type-Options nosniff;
		add_header X-XSS-Protection "1; mode=block";
		add_header Strict-Transport-Security "max-age=31536000; includeSubDomains; preload";
		server_name www.domain.com;
		return 301 https://domain.com$request_uri;
	}

	server {
		listen 443 ssl http2;
		listen [::]:443 ssl http2;
		ssl_certificate /etc/nginx/ssl/domain.com.cert.pem;
		ssl_certificate_key /etc/nginx/ssl/domain.com.key.pem;
		root /var/www/domain.com/html;
		server_name domain.com;
		access_log /var/log/nginx/domain.com.log;
		error_page 404 /404.htm;
		location =/404.htm {etag off;}
		gzip_static on;
		rewrite (.*).gz$ $1 permanent;
		index index.htm;
		if ($request_uri ~ ^(.*/)index.htm$) {return 301 $1;}
		location ~.(png|jpg|ico|cur|css)$ {add_header Cache-Control "public, max-age=31536000, immutable";}
		location ~.exe$ {add_header Cache-Control no-cache;}
		location ~.htm$ {
			add_header Cache-Control no-cache;
			add_header X-Frame-Options SAMEORIGIN;
			add_header X-Content-Type-Options nosniff;
			add_header X-XSS-Protection "1; mode=block";
			add_header Strict-Transport-Security "max-age=31536000; includeSubDomains; preload";
		}
		location =/version.txt {add_header Cache-Control no-cache;}
		location =/robots.txt {add_header Cache-Control "public, max-age=604800"; add_header X-Robots-Tag noindex;}
		location =/sitemap.xml {add_header Cache-Control no-cache; add_header X-Robots-Tag noindex;}
		location /remark42/ {
			rewrite /remark42/(.*) /$1 break;
			proxy_pass http://127.0.0.1:8080/;
			proxy_set_header Host $http_host;
			proxy_set_header X-Real-IP $remote_addr;
			proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
			proxy_set_header X-Forwarded-Proto $scheme;
		}
	}
}
```

**Reply by umputun** [reply]
- what is the remark42 configuration you use? i.e. environment vars or command line.
- How this connected to your site, i.e. the frontend side of the configuration? 
- Does it work with domain.com/remark42/web?
- Does it work with http://127.0.0.1:8080/web ?

**Comment by Ivan-Chupin** [comment]
> what is the remark42 configuration you use? i.e. environment vars or command line.

Command line. Start as a systemd service.

> How this connected to your site, i.e. the frontend side of the configuration?

```javascript
<div id="remark42"></div>
<script>
var remark_config = {host: "https://domain.com/remark42", site_id: 'domain.com', url: 'https://domain.com/index.htm'};
!function(e,n){for(var o=0;o<e.length;o++){var r=n.createElement("script"),c=".js",d=n.head||n.body;"noModule"in r?(r.type="module",c=".mjs"):r.async=!0,r.defer=!0,r.src=remark_config.host+"/web/"+e[o]+c,d.appendChild(r)}}

(remark_config.components||["embed"],document);
</script>
```

> Does it work with domain.com/remark42/web?

![Untitled-1](https://user-images.githubusercontent.com/83099323/116819052-c214a680-ab87-11eb-8803-cb1c2453c52d.png)

> Does it work with http://127.0.0.1:8080/web ?

```
curl -I --request GET http://127.0.0.1:8080/web/

HTTP/1.1 200 OK
Accept-Ranges: bytes
App-Name: remark42
App-Version: master-c9841b8-20210418T17:50:26
Author: umputun
Cache-Control: max-age=3600, no-cache
Content-Length: 2967
Content-Type: text/html; charset=utf-8
Etag: "3f3683227368fc37b48721b86b418cdb2ea73c05"
Last-Modified: Sun, 18 Apr 2021 22:52:26 GMT
Vary: Origin
X-Rate-Limit-Duration: 1
X-Rate-Limit-Limit: 20.00
X-Rate-Limit-Request-Remote-Addr: 127.0.0.1:41446
Date: Sun, 02 May 2021 15:35:08 GMT
```

**Comment by Ivan-Chupin** [comment]
Embedding into a real page.

![Untitled-2](https://user-images.githubusercontent.com/83099323/116819836-53d1e300-ab8b-11eb-80dc-ba3ec1b38cb7.png)

**Comment by umputun** [comment]
well, obviously you not even reaching remark42 assets. My guess will be - some other part of your nginx configuration prevented it. Not sure what part, but would suggest making the simplest config, like in docs, and make sure it works. Adding your other part after this piece by piece till it stops may help to figure what the issue is.

**Comment by Ivan-Chupin** [comment]
Using the exclusion method, I found a line in the nginx configuration that causes this effect:

```nginx
location ~.(png|jpg|ico|cur|css)$ {add_header Cache-Control "public, max-age=31536000, immutable";}
```
But I don't understand why this is happening.

**Reply by umputun** [reply]
me neither. maybe nginx stops evaluating rules after the first match? i.e. any attempt to reach css file will be handled by this location rule and never passed to the next (proxy) rule

**Reply by paskal** [reply]
"The first matching expression stops the search and nginx will use this location." ([doc](http://nginx.org/en/docs/http/request_processing.html)): you need to make remark42 location as high as possible in your list to prevent interference with other rules, but AFAIK it won't work if it's not done using regex location.

**Comment by Ivan-Chupin** [comment]
I completely forgot that regular expressions are searched after prefix comparisons. The most optimal solution would be to use the `^~` prefix, which disallows further search in regular expressions. I think that it is in this form that you need to write in the documentation.

```nginx
location ^~ /remark42/ {
	rewrite /remark42/(.*) /$1 break;
	proxy_pass http://127.0.0.1:8080/;
	proxy_set_header Host $http_host;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Proto $scheme;
}
```

**Reply by sangheonhan** [reply]
Thank you so much, I've been looking at a lot of things like this and not finding the error.
